### PR TITLE
topsql: add sql sum execution duration information

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs520 <crazycs520@gmail.com>

### What problem does this PR solve?

Add SQL sum execution duration information into TopSQL.

### What is changed and how it works?

Add a statement execution fsm to trace the execution flow, and record execution duration for different kind of statement execution.

The state transition diagram is shown in here: https://drive.google.com/file/d/1yBemhOeuuEOPewF4q08n33D38SgtrQKy/view?usp=sharing

### Check List

Tests 

- [x] Unit test

Side effects

- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by crazycs520** [comment]
@mornyx @breeswish PTAL

File: executor/adapter.go [hunk 1]
```diff
@@ -1263,7 +1264,10 @@ func (a *ExecStmt) observeStmtBeginForTopSQL() {
 func (a *ExecStmt) observeStmtFinishedForTopSQL() {
 	if vars := a.Ctx.GetSessionVars(); variable.TopSQLEnabled() && vars.StmtStats != nil {
 		sqlDigest, planDigest := a.getSQLPlanDigest()
-		vars.StmtStats.OnExecutionFinished(sqlDigest, planDigest)
+		costTime := time.Since(vars.StartTime) + vars.DurationParse
+		vars.StmtStats.OnExecutionFinished(sqlDigest, planDigest, func(item *stmtstats.StatementStatsItem) {
+			item.SumExecNanoDuration += uint64(costTime.Nanoseconds())
+		})
```
Comment (breeswish) [inline executor/adapter.go:1270]: I think you should write your logic inside `OnExecutionFinished`.

File: util/topsql/stmtstats/stmtstatstest/stmtstats_test.go [hunk 2]
```diff
@@ -118,6 +118,7 @@ func TestExecCount(t *testing.T) {
 			if _, ok := sqlDigests[digest.SQLDigest]; ok {
 				found++
 				assert.Equal(t, uint64(ExecCountPerSQL), item.ExecCount)
+				assert.True(t, item.SumExecNanoDuration > uint64(time.Nanosecond*ExecCountPerSQL))
```
Comment (breeswish) [inline util/topsql/stmtstats/stmtstatstest/stmtstats_test.go:121]: This lower bound is not strong enough. time.Nanosecond is simply impossible to achieve, which means even if you count the time incorrectly this test will pass.

File: executor/adapter.go (hunk 1 above)
Comment (crazycs520) [reply executor/adapter.go]: Great, done.

File: util/topsql/stmtstats/stmtstatstest/stmtstats_test.go (hunk 2 above)
Comment (mornyx) [reply util/topsql/stmtstats/stmtstatstest/stmtstats_test.go:121]: An idea: the upper bound can be obtained by calculating the duration of `tk.MustExec()` (take the max duration in multiple executions), and the lower bound can be set as the sum of some exact partial execution duration such as `parsing time` (how to obtain these times is a problem, which may require adding a lot of code).

Comment (crazycs520) [reply util/topsql/stmtstats/stmtstatstest/stmtstats_test.go]: resolved, PTAL

File: executor/adapter.go [hunk 3]
```diff
@@ -1257,13 +1257,18 @@ func (a *ExecStmt) observeStmtBeginForTopSQL() {
 		vars.StmtStats.OnExecutionBegin(sqlDigest, planDigest)
 		// This is a special logic prepared for TiKV's SQLExecCount.
 		vars.StmtCtx.KvExecCounter = vars.StmtStats.CreateKvExecCounter(sqlDigest, planDigest)
+		failpoint.Inject("mockSleepInExec", func(ms failpoint.Value) {
+			time.Sleep(time.Duration(ms.(int)) * time.Millisecond)
+
+		})
```
Comment (breeswish) [inline executor/adapter.go:1263]: Why not simply use a sql with `select sleep(1)` or something? The failpoint here doesn't prevent you forget to call `observeStmtBeginForTopSQL` or call it in wrong place, result in not very effective test cases.

File: executor/adapter.go [hunk 4]
```diff
@@ -1263,7 +1263,8 @@ func (a *ExecStmt) observeStmtBeginForTopSQL() {
 func (a *ExecStmt) observeStmtFinishedForTopSQL() {
 	if vars := a.Ctx.GetSessionVars(); variable.TopSQLEnabled() && vars.StmtStats != nil {
 		sqlDigest, planDigest := a.getSQLPlanDigest()
-		vars.StmtStats.OnExecutionFinished(sqlDigest, planDigest)
+		execDuration := time.Since(vars.StartTime) + vars.DurationParse
```
[the remark this answers is not in the document]
Comment (breeswish) [reply executor/adapter.go:1270]: How about passing the `ExecStmt` in?

[the remark this answers is not in the document]
Comment (mornyx) [reply executor/adapter.go:1270]: This may lead to circular dependencies:
```
ExecStmt -> SessionVars -> stmtstats.StatementStats
```

[the remark this answers is not in the document]
Comment (breeswish) [reply executor/adapter.go:1270]: @mornyx Indeed. What I'm worrying is that, seems that the parsing stage happens before the execution stage (which is correct) and the SQL duration needs to include the parsing stage time (and optimization stage time). Is it possible for your framework to cover these stages?

[the remark this answers is not in the document]
Comment (mornyx) [reply executor/adapter.go:1270]: Although the execution stage is after the parsing stage, we can also pass the necessary data of the parsing stage directly as fields of `StatementExecutionContext`.

So we can ignore "what stage happened and when", just add the required fields in `StatementExecutionContext`. For example, for SQLDuration, we need to get `StartTime` and `DurationParse` from `SessionVars` and place them in `StatementExecutionContext`, and then we use them to calculate `SQLDuration` internally.

For future, we also only need to simply add fields in `StatementExecutionContext` (even if it is still not elegant enough).

[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/adapter.go]: Great, done.

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -976,6 +976,10 @@ type SessionVars struct {
 	// all the local data in each session, and finally report them to the remote
 	// regularly.
 	StmtStats *stmtstats.StatementStats
+	// StmtExecCtx represents the current statement execution information.
+	// Note that the information here is not updated in a timely，
+	// it is only updated at the time of `StmtStats.OnExecutionBegin` and `StmtStats.OnExecutionFinished`.
+	StmtExecCtx *stmtstats.StatementExecutionContext
```
Comment (mornyx) [inline sessionctx/variable/session.go:982]: I think we can place it in `ExecStmt`, as a private field

File: executor/adapter.go [hunk 6]
```diff
@@ -1254,16 +1255,27 @@ func (a *ExecStmt) GetTextToLog() string {
...
-		vars.StmtStats.OnExecutionFinished(sqlDigest, planDigest)
+		if vars.StmtExecCtx == nil {
+			sqlDigest, planDigest := a.getSQLPlanDigest()
+			vars.StmtExecCtx = &stmtstats.StatementExecutionContext{
+				SQLDigest:  sqlDigest,
+				PlanDigest: planDigest,
+			}
+		}
+		vars.StmtExecCtx.ExecDuration = time.Since(vars.StartTime) + vars.DurationParse
+		vars.StmtStats.OnExecutionFinished(vars.StmtExecCtx)
```
Comment (mornyx) [inline executor/adapter.go:1278]: Should we pass `StartTime` and `DurationParse` in `StatementExecutionContext `? And then calculate `ExecDuration` in the implementation of `OnExecutionFinished` instead of passing the result here.

Comment (crazycs520) [reply executor/adapter.go:1278]: I think `OnExecutionFinished` shouldn't know how to calculate the `ExecDuration`, so I prefer just calculate in here.

File: sessionctx/variable/session.go (hunk 5 above)
Comment (crazycs520) [reply sessionctx/variable/session.go]: great, done.

File: executor/adapter.go (hunk 6 above)
Comment (mornyx) [reply executor/adapter.go]: We have done this, ExecCount is calculated inside `stmtstats`:
```go
func (s *StatementStats) OnExecutionBegin(ctx *StatementExecutionContext) {
	if ctx == nil {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	item := s.GetOrCreateStatementStatsItem(ctx.SQLDigest, ctx.PlanDigest)

	item.ExecCount++
	// Count more data here.
}
```

So this is possible:
```go
func (s *StatementStats) OnExecutionFinished(ctx *StatementExecutionContext) {
	if ctx == nil {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	item := s.GetOrCreateStatementStatsItem(ctx.SQLDigest, ctx.PlanDigest)

	item.SumExecNanoDuration += time.Since(ctx.StartTime) + ctx.DurationParse
	// Count more data here.
}
```

File: util/topsql/stmtstats/stmtstatstest/stmtstats_test.go [hunk 7]
```diff
@@ -80,22 +80,35 @@ func TestExecCount(t *testing.T) {
...
+		sql := fmt.Sprintf("insert into t values (%d, sleep(0.01));", n)
+		if n == 0 {
+			_, insertSQLDigest = parser.NormalizeDigest(sql)
+		}
+		tk.MustExec(sql)
 	}
-	_, updateSQLDigest := parser.NormalizeDigest("update t set a = 0 where a = 0;")
 	for n := 0; n < ExecCountPerSQL; n++ {
-		tk.MustExec(fmt.Sprintf("update t set a = %d where a = %d;", n, n))
+		sql := fmt.Sprintf("update t set a = %d where a = %d and sleep(0.01);", n, n)
```
Comment (breeswish) [inline util/topsql/stmtstats/stmtstatstest/stmtstats_test.go:93]: This sleep duration is only 10ms. It can be very likely that you will get an actual execution duration very different to 10ms and makes your test not effective.

File: executor/adapter.go [hunk 8]
```diff
@@ -1254,16 +1260,27 @@ func (a *ExecStmt) GetTextToLog() string {
...
-		vars.StmtStats.OnExecutionFinished(sqlDigest, planDigest)
+		if a.stmtExecCtx == nil {
+			sqlDigest, planDigest := a.getSQLPlanDigest()
+			a.stmtExecCtx = &stmtstats.StatementExecutionContext{
+				SQLDigest:  sqlDigest,
+				PlanDigest: planDigest,
+			}
+		}
+		a.stmtExecCtx.ExecDuration = time.Since(vars.StartTime) + vars.DurationParse
+		vars.StmtStats.OnExecutionFinished(a.stmtExecCtx)
```
Comment (breeswish) [inline executor/adapter.go:1283]: How about implemented as:

storing a start time at `OnExecutionBegin` in the `StmtStats` locally, and read it at `OnExecutionBegin`?

File: executor/adapter.go [hunk 9]
```diff
@@ -1254,16 +1260,27 @@ func (a *ExecStmt) GetTextToLog() string {
...
-		sqlDigest, planDigest := a.getSQLPlanDigest()
-		vars.StmtStats.OnExecutionFinished(sqlDigest, planDigest)
+		if a.stmtExecCtx == nil {
+			sqlDigest, planDigest := a.getSQLPlanDigest()
+			a.stmtExecCtx = &stmtstats.StatementExecutionContext{
+				SQLDigest:  sqlDigest,
+				PlanDigest: planDigest,
+			}
+		}
+		a.stmtExecCtx.ExecDuration = time.Since(vars.StartTime) + vars.DurationParse
```
Comment (breeswish) [inline executor/adapter.go:1282]: This context modification is really wired. It should not be modified by the framework level.

File: executor/adapter.go [hunk 10]
```diff
@@ -1254,16 +1260,27 @@ func (a *ExecStmt) GetTextToLog() string {
...
 		// This is a special logic prepared for TiKV's SQLExecCount.
 		vars.StmtCtx.KvExecCounter = vars.StmtStats.CreateKvExecCounter(sqlDigest, planDigest)
 	}
 }
 
 func (a *ExecStmt) observeStmtFinishedForTopSQL() {
 	if vars := a.Ctx.GetSessionVars(); variable.TopSQLEnabled() && vars.StmtStats != nil {
-		sqlDigest, planDigest := a.getSQLPlanDigest()
-		vars.StmtStats.OnExecutionFinished(sqlDigest, planDigest)
+		if a.stmtExecCtx == nil {
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/adapter.go:1275]: Normally it should never happen, but if some unknown statement that doesn't go through `observeStmtBeginForTopSQL`, it will happen.

File: util/topsql/stmtstats/stmtstatstest/stmtstats_test.go (hunk 7 above)
Comment (crazycs520) [reply util/topsql/stmtstats/stmtstatstest/stmtstats_test.go:93]: Ok. But in my laptop, those SQLs execution duration is around 11ms, and then subtract sleep time by 10ms, the SQL itself only execute about 1ms.  I will enlarge this sleep for machines with different performance.

File: util/topsql/stmtstats/stmtstats.go [hunk 11]
```diff
@@ -16,30 +16,47 @@ package stmtstats
...
-	// OnExecutionBegin should be called before statement execution.
-	OnExecutionBegin(sqlDigest, planDigest []byte)
+	// OnParseBegin should be called on statement begin to parse.
+	OnParseBegin()
+
+	// OnSQLDigestReady should be called on statement SQL digest ready.
+	OnSQLDigestReady(sqlDigest []byte)
+
+	// OnPlanDigestReady should be called on statement plan digest ready.
+	OnPlanDigestReady(sqlDigest []byte)
```
Comment (mornyx) [inline util/topsql/stmtstats/stmtstats.go:42]:
```suggestion
	OnPlanDigestReady(planDigest []byte)
```

File: util/topsql/stmtstats/stmtstats.go [hunk 12]
```diff
@@ -55,19 +72,57 @@ func CreateStatementStats() *StatementStats {
...
+
+// OnExecBegin implements StatementObserver.OnExecBegin.
+func (s *StatementStats) OnExecBegin() {
+	s.seCtx.execBeginUnixNano = nowFunc().UnixNano()
+}
+
+// OnExecFinished implements StatementObserver.OnExecFinished.
+func (s *StatementStats) OnExecFinished() {
+	sqlDigest, ok1 := s.seCtx.sqlDigest.Load().([]byte)
+	planDigest, ok2 := s.seCtx.planDigest.Load().([]byte)
```
Comment (mornyx) [inline util/topsql/stmtstats/stmtstats.go:98]: If `planDigest` is nil, will the `.([]byte)` panic here?

File: server/conn_stmt.go [hunk 13]
```diff
@@ -280,6 +285,8 @@ func (cc *clientConn) handleStmtFetch(ctx context.Context, data []byte) (err err
 		prepareObj, _ := cc.preparedStmtID2CachePreparedStmt(stmtID)
 		if prepareObj != nil && prepareObj.SQLDigest != nil {
 			ctx = topsql.AttachSQLInfo(ctx, prepareObj.NormalizedSQL, prepareObj.SQLDigest, "", nil, false)
+			sessVars.StmtStats.OnSQLDigestReady(prepareObj.SQLDigest.Bytes())
```
Comment (mornyx) [inline server/conn_stmt.go:288]: It seems that `OnSQLDigestReady` will be called multiple times in the same execution. Is this expected? If so, I suggest to describe it in the comment of `StatementObserver.OnSQLDigestReady` that this method may be called multiple times and must be no side effects within a single execution.

Comment (crazycs520) [reply server/conn_stmt.go:288]: Already add some comment. `OnSQLDigestReady` is ok to call multiple times.
But currently, `OnSQLDigestReady` won't call multiple times for 1 statement. here is a special handle for `stmt fetch` statement.

File: util/topsql/stmtstats/stmtstats.go (hunk 12 above)
Comment (crazycs520) [reply util/topsql/stmtstats/stmtstats.go:105]: `seCtx.planDigest` type is `atomic.Value`, it won't be nil. if the value of `seCtx.planDigest` stored is nil, it won't be panic here, but the `ok2` will be false.  Also there are some test call `OnPlanDigestReady(nil)` to store nil in `stmtstats_test.go`

File: util/topsql/stmtstats/stmtstats.go [hunk 14]
```diff
@@ -16,30 +16,54 @@ package stmtstats
...
+
 // StatementObserver is an abstract interface as a callback to the corresponding
 // position of TiDB's SQL statement execution process. StatementStats implements
 // StatementObserver and performs counting such as SQLExecCount/SQLDuration internally.
 // The caller only needs to be responsible for calling different methods at the
 // corresponding locations, without paying attention to implementation details.
 type StatementObserver interface {
-	// OnExecutionBegin should be called before statement execution.
-	OnExecutionBegin(sqlDigest, planDigest []byte)
+	// OnParseBegin should be called on statement begin to parse.
```
Comment (crazycs520) [inline util/topsql/stmtstats/stmtstats.go:35]: OnRecieveCmd()

Comment (crazycs520) [reply util/topsql/stmtstats/stmtstats.go:35]: state: initial -> cmd recieve (reset state)  -> sql/plan digest raedy -> exec begin -> finish     , if state not right, log it out.

File: util/topsql/stmtstats/stmtstats.go [hunk 15]
```diff
@@ -55,19 +79,57 @@ func CreateStatementStats() *StatementStats {
...
+		item.SumExecNanoDuration += uint64(cost)
+	}
 	// Count more data here.
-}
 
-// OnExecutionFinished implements StatementObserver.OnExecutionFinished.
-func (s *StatementStats) OnExecutionFinished(sqlDigest, planDigest []byte) {
-	// Count more data here.
+	// Reset StatementExecutionContext for next statement.
+	s.seCtx.reset()
```
Comment (crazycs520) [inline util/topsql/stmtstats/stmtstats.go:132]: move to on cmd reveived.

File: util/topsql/stmtstats/stmtstats.go [hunk 16]
```diff
@@ -55,19 +79,57 @@ func CreateStatementStats() *StatementStats {
...
 
-// OnExecutionBegin implements StatementObserver.OnExecutionBegin.
-func (s *StatementStats) OnExecutionBegin(sqlDigest, planDigest []byte) {
+// OnParseBegin implements StatementObserver.OnParseBegin.
+func (s *StatementStats) OnParseBegin() {
+	s.seCtx.parseBeginUnixNano = nowFunc().UnixNano()
+}
+
+// OnSQLDigestReady implements StatementObserver.OnSQLDigestReady.
+func (s *StatementStats) OnSQLDigestReady(sqlDigest []byte) {
```
Comment (crazycs520) [inline util/topsql/stmtstats/stmtstats.go:88]: OnSQLPlanDigestReady()   //

**Comment by breeswish** [comment]
1. Events:

- OnReceiveCmd
- OnSqlAndPlanDigestFirstReady
- OnExecFinish

2. Add state check:

INITIAL, CMD_RECEIVED, DIGEST_IS_READY, FINISHED

3. In OnSqlAndPlanDigestFirstReady: Exec Count +1

4. In OnExecFinish: Add duration

5. In OnReceiveCmd: Store initial ts (in `time.Time` to resist time going back)

6. Context: No atomic is needed

File: util/topsql/stmtstats/stmtstats.go [hunk 17]
```diff
@@ -15,59 +15,134 @@
...
 	mu       sync.Mutex
 	data     StatementStatsMap
 	finished *atomic.Bool
 }
 
+// StmtExecState is the state of StatementStats
+type StmtExecState int
+
+var (
+	stmtExecStateInitial       StmtExecState = 0
```
Comment (breeswish) [inline util/topsql/stmtstats/stmtstats.go:65]:
```suggestion
	stmtExecStateInitial       StmtExecState = iota
```

File: util/topsql/stmtstats/stmtstats.go [hunk 18]
```diff
@@ -15,59 +15,134 @@
...
+		return "initial"
+	case stmtExecStateCmdReceived:
+		return "cmd_received"
+	case stmtExecStateDigestIsReady:
+		return "digest_is_ready"
+	case stmtExecStateFinished:
+		return "finish"
+	default:
+		return fmt.Sprintf("unknow_%d", int(s))
+	}
```
Comment (breeswish) [inline util/topsql/stmtstats/stmtstats.go:84]: If you want to pretty print it, then how about type it as a string, like `type StmtExecState string`?

File: util/topsql/stmtstats/stmtstats.go [hunk 19]
```diff
@@ -15,59 +15,134 @@
...
+// OnReceiveCmd implements StatementObserver.OnParseBegin.
+func (s *StatementStats) OnReceiveCmd() {
+	s.state = stmtExecStateCmdReceived
+	s.seCtx.reset()
+}
+
+// OnSQLAndPlanDigestFirstReady implements StatementObserver.OnSQLAndPlanDigestFirstReady.
+func (s *StatementStats) OnSQLAndPlanDigestFirstReady(sqlDigest, planDigest []byte) {
+	if s.state != stmtExecStateCmdReceived {
+		logutil.BgLogger().Warn("[stmt-stats] unexpect state",
```
Comment (breeswish) [inline util/topsql/stmtstats/stmtstats.go:107]:
```suggestion
		logutil.BgLogger().Warn("[stmt-stats] unexpected state",
```

File: util/topsql/stmtstats/stmtstats.go [hunk 20]
```diff
@@ -15,59 +15,134 @@
...
+			zap.String("expect", stmtExecStateDigestIsReady.String()),
+			zap.String("got", s.state.String()))
+		return
+	}
+	s.state = stmtExecStateFinished
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if len(s.seCtx.sqlDigest) == 0 {
```
Comment (breeswish) [inline util/topsql/stmtstats/stmtstats.go:137]: this should never happen?

File: util/topsql/stmtstats/stmtstats.go [hunk 21]
```diff
@@ -15,59 +15,134 @@
...
 		finished: atomic.NewBool(false),
 	}
 	globalAggregator.register(stats)
 	return stats
 }
 
-// OnExecutionBegin implements StatementObserver.OnExecutionBegin.
-func (s *StatementStats) OnExecutionBegin(sqlDigest, planDigest []byte) {
+// OnReceiveCmd implements StatementObserver.OnParseBegin.
+func (s *StatementStats) OnReceiveCmd() {
```
Comment (breeswish) [inline util/topsql/stmtstats/stmtstats.go:99]: maybe better to check and output invalid state here, for example, it expects state to be initial or finished (but proceed with overridden state)

**Review by breeswish — COMMENTED** [review]
the rest LGTM

File: util/topsql/stmtstats/stmtstats.go (hunk 18 above)
Comment (crazycs520) [reply util/topsql/stmtstats/stmtstats.go:84]: I prefer use int here, since execute compare `int` is faster than compare `string`.

Comment (breeswish) [reply util/topsql/stmtstats/stmtstats.go:84]: I don't think it is really necessary for this trivial optimization. It only reduce less than 1ns. You will never get real performance improvements by focusing on these non performance critical paths. If you really care about the performance, then reduce your allocations, improve your cache efficiency, reduce atomic contentions.

Comment (crazycs520) [reply util/topsql/stmtstats/stmtstats.go]: Ok, but I still prefer using `int` here.

File: util/topsql/stmtstats/state.go [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+		if *s == state {
+			found = true
+			break
+		}
+	}
+	if found {
+		*s = target
+		return true
+	}
+	logutil.BgLogger().Warn("[stmt-stats] unexpected state", zap.String("state", s.String()), zap.String("target", target.String()))
```
Comment (crazycs520) [inline util/topsql/stmtstats/state.go:46]: TODO: should change this log level to debug.

File: util/topsql/stmtstats/stmtstats.go [hunk 23]
```diff
@@ -16,30 +16,75 @@ package stmtstats
...
 // StatementObserver is an abstract interface as a callback to the corresponding
 // position of TiDB's SQL statement execution process. StatementStats implements
 // StatementObserver and performs counting such as SQLExecCount/SQLDuration internally.
 // The caller only needs to be responsible for calling different methods at the
 // corresponding locations, without paying attention to implementation details.
 type StatementObserver interface {
-	// OnExecutionBegin should be called before statement execution.
-	OnExecutionBegin(sqlDigest, planDigest []byte)
+	// OnDispatchBegin should be called on the begin of TiDB dispatch command.
+	OnDispatchBegin()
```
Comment (breeswish) [inline util/topsql/stmtstats/stmtstats.go:36]: I think you should explicitly point out protocol level things and other things. For protocol level things, I prefer to add a Cmd prefix, since Command is a unique concept that will not cause ambiguous:

- OnCmdDispatchBegin
- OnCmdDispatchFinish
- OnCmdQueryBegin
- OnCmdQueryFinish
- OnCmdQueryProcessStmtBegin
- OnCmdQueryProcessStmtFinish
- OnCmdStmtExecuteBegin
- OnCmdStmtExecuteFinish
- OnCmdStmtFetchBegin
- OnCmdStmtFetchFinish
- OnStmtReadyToExecute

This can be very clear that where these events should be called.

File: util/topsql/stmtstats/stmtstats.go [hunk 24]
```diff
@@ -16,30 +16,75 @@ package stmtstats
...
+	OnUseDBBegin()
+
+	// OnHandleStmtFinish should be called on TiDB finished use db.
+	OnUseDBFinish()
+
+	// OnHandleInternalStmtBegin should be called on the begin of TiDB handle internal stmt.
+	OnHandleInternalStmtBegin()
+
+	// OnHandleInternalStmtFinish should be called on TiDB finished handle internal stmt.
+	OnHandleInternalStmtFinish()
```
Comment (breeswish) [inline util/topsql/stmtstats/stmtstats.go:79]: Why do we need an internal statement? What will happen if it is not processed? Will we miss background SQL calls?

File: go.mod [hunk 25]
```diff
@@ -10,6 +10,7 @@ require (
 	github.com/Jeffail/gabs/v2 v2.5.1
 	github.com/aws/aws-sdk-go v1.35.3
 	github.com/blacktear23/go-proxyprotocol v0.0.0-20180807104634-af7a81e8dd0d
+	github.com/breeswish/go-litefsm v0.0.0-20211228123400-53ff62254a35 // indirect
```
Comment (breeswish) [inline go.mod:13]: you need to use `go mod tidy` to make it indirect.

File: server/conn.go [hunk 26]
```diff
@@ -1255,6 +1255,9 @@ func (cc *clientConn) dispatch(ctx context.Context, data []byte) error {
 	cc.lastPacket = data
 	cmd := data[0]
 	data = data[1:]
+	vars := cc.ctx.GetSessionVars()
```
Comment (breeswish) [inline server/conn.go:1258]: Do you need to check whether vars is nil and vars.StmtStats is nil?

**Review by breeswish — COMMENTED** [review]
I'm worried about whether there will be race when calling OnExecuteBegin/Finish and OnDigestKnown. The rest looks fine.

File: server/conn.go (hunk 26 above)
Comment (crazycs520) [reply server/conn.go:1258]: Look like they should never be nil, so no need to check it?

File: executor/adapter.go [hunk 27]
```diff
@@ -1283,22 +1282,15 @@ func (a *ExecStmt) GetTextToLog() string {
 	return sql
 }
 
-func (a *ExecStmt) observeStmtBeginForTopSQL() {
+func (a *ExecStmt) observeOnDigestKnownForTopSQL() {
 	if vars := a.Ctx.GetSessionVars(); variable.TopSQLEnabled() && vars.StmtStats != nil {
```
Comment (breeswish) [inline executor/adapter.go:1286]: When there is a session var, `vars.StmtStats` must not be nil.

File: executor/adapter.go [hunk 28]
```diff
@@ -1283,22 +1282,15 @@ func (a *ExecStmt) GetTextToLog() string {
...
 		sqlDigest, planDigest := a.getSQLPlanDigest()
-		vars.StmtStats.OnExecutionBegin(sqlDigest, planDigest)
+		vars.StmtStats.OnDigestKnown(sqlDigest, planDigest)
 		// This is a special logic prepared for TiKV's SQLExecCount.
 		vars.StmtCtx.KvExecCounter = vars.StmtStats.CreateKvExecCounter(sqlDigest, planDigest)
```
Comment (breeswish) [inline executor/adapter.go:1290]: Is it possible to move this into `OnDigestKnown`?

Comment (crazycs520) [reply executor/adapter.go:1290]: No, since `KvExecCounter ` is in `StmtCtx ` , not `StmtStats `

Comment (breeswish) [reply executor/adapter.go:1290]: I mean is it posssible to move it from `StmtCtx` into `vars.StmtStats`? As you can see, this PR did a hard work to maintain the semantic of "execution", and looks like it is reliable enough now for KvExecCounter to be placed inside.

File: util/topsql/stmtstats/observer.go [hunk 29]
```diff
@@ -39,7 +39,7 @@ type StatementObserver interface {
 	// It may be called without a MySQL protocol when this session is created in background.
 	OnDigestKnown(sqlDigest []byte, planDigest []byte)
 
-	// The following events should be called during reading from wire protocol.
-	OnExecuteBegin()
-	OnExecuteFinish()
+	// The following events should be called during reading from internal protocol.
```
Comment (breeswish) [inline util/topsql/stmtstats/observer.go:42]: There is no internal protocol.

```suggestion
	// The following events should be called when executing an internal statement. The execution may happen
	// directly over a session that is built from the wire connection, or an internal session in the pool.
```

**Comment by crazycs520** [comment]
see https://github.com/pingcap/tidb/pull/31143
